platforms/linux/components.c

```diff
@@ -0,0 +1,496 @@
+                           unsigned long arg)
+{
+    int ret = 0;
+    void *argp = (void*)arg;
+    struct vcpu_t *cvcpu;
+    struct hax_vcpu_linux_t *vcpu;
+    struct miscdevice *miscdev;
+
+    miscdev = filp->private_data;
+    vcpu = container_of(miscdev, struct hax_vcpu_linux_t, dev);
+    cvcpu = hax_get_vcpu(vcpu->vm->id, vcpu->id, 1);
+    if (!cvcpu)
+        return -ENODEV;
+
+    switch (cmd) {
+    case HAX_VCPU_IOCTL_RUN:
+        ret = vcpu_execute(cvcpu);
+        break;
+    case HAX_VCPU_IOCTL_SETUP_TUNNEL: {
+        struct hax_tunnel_info info;
+        ret = hax_vcpu_setup_hax_tunnel(cvcpu, &info);
+        if (copy_to_user(argp, &info, sizeof(info))) {
+            ret = -EFAULT;
+            break;
+        }
+        break;
+    }
+    case HAX_VCPU_IOCTL_SET_MSRS: {
+        struct hax_msr_data msrs;
+        struct vmx_msr *msr;
+        int i, fail;
+
+        if (copy_from_user(&msrs, argp, sizeof(msrs))) {
+            ret = -EFAULT;
+            break;
+        }
+        msr = msrs.entries;
+        /* nr_msr needs to be verified */
+        if (msrs.nr_msr >= 0x20) {
+            hax_error("MSRS invalid!\n");
+            return -EFAULT;
+        }
+        for (i = 0; i < msrs.nr_msr; i++, msr++) {
+            fail = vcpu_set_msr(cvcpu, msr->entry, msr->value);
+            if (fail) {
+                break;
+            }
+        }
+        msrs.done = i;
+        break;
+    }
+    case HAX_VCPU_IOCTL_GET_MSRS: {
+        struct hax_msr_data msrs;
+        struct vmx_msr *msr;
+        int i, fail;
+
+        if (copy_from_user(&msrs, argp, sizeof(msrs))) {
+            ret = -EFAULT;
+            break;
+        }
+        msr = msrs.entries;
+        if(msrs.nr_msr >= 0x20) {
+            hax_error("MSRS invalid!\n");
+            return -EFAULT;
+        }
+        for (i = 0; i < msrs.nr_msr; i++, msr++) {
+            fail = vcpu_get_msr(cvcpu, msr->entry, &msr->value);
+            if (fail) {
+                break;
+            }
+        }
+        msrs.done = i;
+        if (copy_to_user(argp, &msrs, sizeof(msrs))) {
+            ret = -EFAULT;
+            break;
+        }
+        break;
+    }
+    case HAX_VCPU_IOCTL_SET_FPU: {
+        struct fx_layout fl;
+        if (copy_from_user(&fl, argp, sizeof(fl))) {
+            ret = -EFAULT;
+            break;
+        }
+        ret = vcpu_put_fpu(cvcpu, &fl);
+        break;
+    }
+    case HAX_VCPU_IOCTL_GET_FPU: {
+        struct fx_layout fl;
+        ret = vcpu_get_fpu(cvcpu, &fl);
+        if (copy_to_user(argp, &fl, sizeof(fl))) {
+            ret = -EFAULT;
+            break;
+        }
+        break;
+    }
+    case HAX_VCPU_SET_REGS: {
+        struct vcpu_state_t vc_state;
+        if (copy_from_user(&vc_state, argp, sizeof(vc_state))) {
+            ret = -EFAULT;
+            break;
+        }
+        ret = vcpu_set_regs(cvcpu, &vc_state);
+        // Due to debugging, vcpu_set_regs might modify drN registers
```

Comment:
Just realized that I had referred you to the wrong QEMU-side patch... @junxiaoc will delete it and re-upload the correct one. Sorry!

Reply:
@AlexAltea This issue is not addressed by https://github.com/intel/haxm/pull/108/commits/ac265fd651b00f8c9e6ee9a8a808781d5f203d96, so I can't approve the PR yet. What's your take on it?

BTW, @junxiaoc has updated #81 with the correct QEMU patch.

Reply:
Sorry, I had posponed this issue and ended up forgetting about it.

The reason I added `copy_to_user` was because I saw. But I agree this is completely counter-intuitive, since `SET_REGS` should never change the userland-provided struct. I'll remove those lines.

Reply:
> The reason I added copy_to_user was because I saw vcpu_set_regs modifying the state.

Ah, you must be referring to this part:

```
        ustate->_dr6 = fix_dr6(ustate->_dr6);
        UPDATE_VCPU_STATE(_dr6, dr_dirty);
        ustate->_dr7 = fix_dr7(ustate->_dr7);
        UPDATE_VCPU_STATE(_dr7, dr_dirty);
```

The purpose of `fix_dr6/fix_dr7` is to make sure we correctly set or clear all the reserved bits. Ideally we shouldn't modify the input `ustate`, but ended up doing that so we could keep using the `UPDATE_VCPU_STATE` macro for DR6 and DR7 as before.